mlflow/llama_index/__init__.py

```diff
@@ -0,0 +1,385 @@
+        path: Local path where the serialized model (as YAML) is to be saved.
+        engine_type: Determine how the instantiated index can be interacted with. The supported
+            types are as follows:
+
+            - "chat": instantiate the object as a llama_index ``chat_engine``.
+            - "query": instantiate the object as a llama_index ``query_engine``.
+            - "retriever": instantiate the object as a llama_index ``retriever``.
+
+        engine_kwargs: Keyword arguments to be passed to the llama_index engine at instantiation.
+            Note that not all llama index objects have supported serialization; when an object is
+            not supported, an info log message will be emitted and the unsupported object will be
+            dropped.
+        code_paths: {{ code_paths }}
+        mlflow_model: An MLflow model object that specifies the flavor that this model is being
+            added to.
+        signature: A Model Signature object that describes the input and output Schema of the
+            model. The model signature can be inferred using `infer_signature` function
+            of `mlflow.models.signature`.
+        input_example: {{ input_example }}
+        pip_requirements: {{ pip_requirements }}
+        extra_pip_requirements: {{ extra_pip_requirements }}
+        conda_env: {{ conda_env }}
+        metadata: {{ metadata }}
+        : Optional additional configurations for transformers serialization.
+    """
+    import llama_index
+
+    if engine_type not in _SUPPORTED_ENGINES:
+        raise ValueError(
+            f"Currently mlflow only supports the following engine types: "
+            f"{_SUPPORTED_ENGINES}. {engine_type} is not supported, so please"
+            "use one of the above types"
+        )
+
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+
+    path = os.path.abspath(path)
+    _validate_and_prepare_target_save_path(path)
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+
+    if signature is None and input_example is not None:
+        wrapped_model = _LlamaIndexModelWrapper(index)
+        signature = infer_signature_from_input_example(input_example, wrapped_model)
+    elif signature is False:
+        signature = None
+
+    if mlflow_model is None:
+        mlflow_model = Model()
+    if signature is not None:
+        signature = validate_and_resolve_signature(index, signature)
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+    if metadata is not None:
+        mlflow_model.metadata = metadata
+
+    engine_kwargs_path = os.path.join(path, _ENGINE_KWARGS_DIRECTORY)
+    serialize_engine_kwargs_to_json(engine_kwargs, engine_kwargs_path)
+
+    from llama_index.core import Settings
+
+    settings_path = os.path.join(path, _SETTINGS_DIRECTORY)
+    serialize_settings_to_json(Settings, settings_path)
```

Comment:
This needs a dev note comment explaining what is happening here. Why we're accessing a singleton and what we're extracting from it.